File: qa-automation/AI-Scoring/backend/services/onepager.py
"""Agent one-pager rendering — the print-ready monthly HTML (JulyR2R3 §3).

Moved verbatim from ``scripts/export_onepager.py`` so the dashboard can
serve the same artifact on demand; the script stays the CLI wrapper for
the operator's monthly file export (and is the only caller that
*generates* assessments — see ``render_month_onepager``).

Two entry layers:

- Pure render pieces (``frame_to_render`` … ``render``): dataframe slice
  in, standalone HTML out. No I/O, unit-testable.
- ``render_month_onepager``: the route-facing orchestrator — fetches the
  analytics frame (same row-source + bucket-TZ month semantics as the
  dashboard), slices one agent+month, attaches the *persisted* month
  assessment (``generate=False`` default: a page view must never spend a
  Gemini call), and renders. Returns None when the agent has no rows in
  the month.
"""

from __future__ import annotations

import html
import statistics
from datetime import datetime
from zoneinfo import ZoneInfo

import pandas as pd

from backend.config.team_config import TeamConfig
from backend.services.team_stats import BUCKET_TZ_NAME

_BUCKET_TZ = ZoneInfo(BUCKET_TZ_NAME)

# Landing brand palette (qa-automation/teams/*/Branding.js)
NAVY = "#15192D"
BLUE = "#1A61D9"
LIGHT_BLUE = "#E7EFFB"
GREEN = "#28A745"
AMBER = "#E8A317"
RED = "#D9534F"
GRAY = "#4A4A4A"

_YN_VALUES = {"Yes": 1.0, "Y": 1.0, "No": 0.0, "N": 0.0}


def last_closed_month(now: datetime | None = None) -> str:
    """``YYYY-MM`` of the most recently completed bucket-TZ month — what
    the dashboard link and the monthly exports both mean by "the month":
    on the 1st it rolls forward together with the workbook export."""
    local = (now or datetime.now(_BUCKET_TZ)).astimezone(_BUCKET_TZ)
    if local.month == 1:
        return f"{local.year - 1}-12"
    return f"{local.year}-{local.month - 1:02d}"


def frame_to_render(df_month: pd.DataFrame, agent: str, config):
    """One agent's month slice of the analytics frame → the render shape
    (display-label columns, canonical section order). The frame's yn cells
    are already the Y/N/NA/'' vocabulary _section_stats understands;
    numeric NAs arrive as NaN (the frame collapses the sheet-era
    'Not Applicable' string — the NA count column reflects yn sections
    only, a known fidelity trade)."""
    g = df_month[df_month["agent"] == agent].sort_values("timestamp")
    out = pd.DataFrame({
        "Agent Name": g["agent"].values,
        "_ts": g["timestamp"].values,
        "Overall Score": g["overall_score"].values,
    })
    section_cols = []
    for sec in config.sections_by_number:
        h = sec.history_id
        if h in g.columns:
            out[sec.name] = g[h].values
            section_cols.append(sec.name)
    return out, section_cols


def assessment_html(assessment: dict | None) -> str:
    """The AI-Assessment slot: the persisted qa.assessments row, or the
    reserved-slot placeholder when none exists."""
    if assessment is None:
        return (
            '<div class="assessment"><strong>AI Assessment</strong><br>'
            '<span class="muted">No persisted assessment for this window — '
            "the monthly export run creates one.</span></div>"
        )
    lines = []
    for s in assessment.get("sections", []):
        arrow = {"improving": ("▲", GREEN), "declining": ("▼", RED)}.get(
            s["trend"], ("→", GRAY))
        lines.append(
            f'<div class="a-sec"><span style="color:{arrow[1]}">{arrow[0]}</span> '
            f"<strong>{html.escape(s['section_name'])}</strong> — "
            f"{html.escape(s['coaching_tip'])}</div>"
        )
    gen = assessment.get("generated_at")
    stamp = gen.strftime("%Y-%m-%d") if hasattr(gen, "strftime") else ""
    return (
        '<div class="assessment"><strong>AI Assessment</strong> '
        f'<span class="muted">({assessment.get("evaluations_included", "?")} evaluations'
        f'{" · " + stamp if stamp else ""} · {html.escape(str(assessment.get("rubric_version", "")))})</span>'
        f'<p style="margin:6px 0 8px">{html.escape(assessment["overall_assessment"])}</p>'
        f"{''.join(lines)}</div>"
    )


def _section_stats(df: pd.DataFrame, col: str) -> dict:
    values = df[col].astype(str).str.strip()
    numeric = pd.to_numeric(values, errors="coerce")
    is_binary = values.isin(_YN_VALUES).any()
    midpoint = len(df) // 2
    out = {"name": col, "na": int(values.isin(["Not Applicable", "NA"]).sum())}
    if is_binary:
        mapped = values.map(_YN_VALUES).dropna()
        out["kind"] = "binary"
        out["avg"] = round(mapped.mean() * 100, 1) if len(mapped) else None
        first, second = mapped.iloc[:midpoint], mapped.iloc[midpoint:]
        out["delta"] = (
            round((second.mean() - first.mean()) * 100, 1)
            if len(first) and len(second) else None
        )
    else:
        scores = numeric.dropna()
        out["kind"] = "numeric"
        out["avg"] = round(scores.mean(), 2) if len(scores) else None
        first, second = scores.iloc[:midpoint], scores.iloc[midpoint:]
        out["delta"] = (
            round(second.mean() - first.mean(), 2)
            if len(first) and len(second) else None
        )
    return out


def _trend_arrow(delta) -> tuple[str, str]:
    if delta is None:
        return "—", GRAY
    if delta > 0.05:
        return f"▲ +{delta:g}", GREEN
    if delta < -0.05:
        return f"▼ {delta:g}", RED
    return "→ steady", GRAY


def _sparkline(scores: list[float], width=560, height=64) -> str:
    """Inline SVG polyline of per-call overall scores."""
    if len(scores) < 2:
        return ""
    lo, hi = min(scores + [60]), max(scores + [100])
    span = (hi - lo) or 1
    step = width / (len(scores) - 1)
    points = [
        (round(i * step, 1), round(height - (s - lo) / span * (height - 8) - 4, 1))
        for i, s in enumerate(scores)
    ]
    polyline = " ".join(f"{x},{y}" for x, y in points)
    dots = "".join(
        f'<circle cx="{x}" cy="{y}" r="2.5" fill="{BLUE}"/>' for x, y in points
    )
    return (
        f'<svg width="{width}" height="{height}" viewBox="0 0 {width} {height}">'
        f'<polyline points="{polyline}" fill="none" stroke="{BLUE}" stroke-width="2"/>'
        f"{dots}</svg>"
    )


def _bar(stat: dict) -> str:
    if stat["avg"] is None:
        return ""
    pct = stat["avg"] if stat["kind"] == "binary" else stat["avg"] / 5 * 100
    color = GREEN if pct >= 80 else AMBER if pct >= 60 else RED
    return (
        f'<div class="bar"><div class="fill" '
        f'style="width:{pct:.0f}%;background:{color}"></div></div>'
    )


def render(df: pd.DataFrame, agent: str, month: str, section_cols: list[str],
           assessment: dict | None = None,
           team_label: str = "Member Support") -> str:
    overall = pd.to_numeric(df["Overall Score"], errors="coerce").dropna().tolist()
    n = len(overall)
    mean = statistics.fmean(overall)
    std = statistics.stdev(overall) if n > 1 else 0.0
    month_label = datetime.strptime(month, "%Y-%m").strftime("%B %Y")
    stats = [_section_stats(df, c) for c in section_cols]

    rows = []
    for s in stats:
        arrow, color = _trend_arrow(s["delta"])
        if s["avg"] is None:
            avg_cell = "—"
        elif s["kind"] == "binary":
            avg_cell = f"{s['avg']:g}% Yes"
        else:
            avg_cell = f"{s['avg']:g} / 5"
        na_cell = f'{s["na"]} NA' if s["na"] else ""
        rows.append(
            f"<tr><td>{html.escape(s['name'])}</td>"
            f'<td class="num">{avg_cell}</td>'
            f"<td>{_bar(s)}</td>"
            f'<td class="num" style="color:{color}">{arrow}</td>'
            f'<td class="num muted">{na_cell}</td></tr>'
        )

    return f"""<!DOCTYPE html>
<html><head><meta charset="utf-8"><title>{html.escape(agent)} — QA {month_label}</title>
<style>
  @page {{ size: letter portrait; margin: 0.55in; }}
  * {{ box-sizing: border-box; margin: 0; }}
  body {{ font: 13px/1.45 -apple-system, "Segoe UI", Helvetica, Arial, sans-serif;
         color: {GRAY}; max-width: 7.4in; margin: 0 auto; padding: 24px 8px; }}
  header {{ background: {NAVY}; color: #fff; border-radius: 10px;
            padding: 18px 22px; display: flex; justify-content: space-between;
            align-items: baseline; }}
  header h1 {{ font-size: 21px; font-weight: 650; }}
  header .brand {{ color: {LIGHT_BLUE}; font-size: 12px; letter-spacing: 0.12em;
                   text-transform: uppercase; }}
  .cards {{ display: grid; grid-template-columns: repeat(4, 1fr); gap: 10px;
            margin: 14px 0; }}
  .card {{ background: {LIGHT_BLUE}; border-radius: 8px; padding: 12px 14px; }}
  .card .v {{ font-size: 26px; font-weight: 700; color: {NAVY};
              font-variant-numeric: tabular-nums; }}
  .card .l {{ font-size: 10.5px; text-transform: uppercase; letter-spacing: 0.08em; }}
  h2 {{ font-size: 12px; color: {NAVY}; text-transform: uppercase;
        letter-spacing: 0.1em; margin: 18px 0 8px; }}
  table {{ width: 100%; border-collapse: collapse; }}
  td {{ padding: 5px 8px; border-bottom: 1px solid #eceff4; }}
  td.num {{ text-align: right; white-space: nowrap;
            font-variant-numeric: tabular-nums; }}
  .muted {{ color: #9aa3af; font-size: 11px; }}
  .bar {{ width: 130px; height: 7px; background: #e8ebf0; border-radius: 4px; }}
  .fill {{ height: 100%; border-radius: 4px; }}
  .assessment {{ border: 1.5px dashed {BLUE}; border-radius: 10px; padding: 14px 16px;
                 margin-top: 16px; color: {NAVY}; background: #fbfcff; }}
  .a-sec {{ font-size: 11.5px; margin: 2px 0; }}
  footer {{ margin-top: 18px; font-size: 10.5px; color: #9aa3af; }}
</style></head><body>
<header>
  <div><h1>{html.escape(agent)}</h1>
  <div style="color:{LIGHT_BLUE}">{html.escape(team_label)} — QA Performance</div></div>
  <div style="text-align:right"><div class="brand">Landing · QA</div>
  <div style="font-size:15px;font-weight:600">{month_label}</div></div>
</header>

<div class="cards">
  <div class="card"><div class="v">{mean:.1f}</div><div class="l">Avg overall score</div></div>
  <div class="card"><div class="v">{n}</div><div class="l">Calls evaluated</div></div>
  <div class="card"><div class="v">{std:.1f}</div><div class="l">Std deviation</div></div>
  <div class="card"><div class="v">{min(overall):.0f}–{max(overall):.0f}</div><div class="l">Score range</div></div>
</div>

<h2>Score progression — {month_label}</h2>
{_sparkline(overall)}

<h2>Per-section performance</h2>
<table><tbody>
{chr(10).join(rows)}
</tbody></table>

{assessment_html(assessment)}

<footer>Generated {datetime.now().strftime("%Y-%m-%d %H:%M")} · Source: qa.evaluations
({n} finalized evaluations, {month_label}) · Trend = second-half vs first-half average
· Scores computed by the Landing QA engine</footer>
</body></html>"""


async def render_month_onepager(
    config: TeamConfig,
    agent: str,
    month: str,
    *,
    generate: bool = False,
) -> str | None:
    """Fetch, slice, and render one agent's month one-pager.

    ``generate=False`` (the route default) never spends a Gemini call: the
    assessment slot shows whatever the monthly export run persisted, or
    the placeholder. Only the operator CLI passes ``generate=True``.
    Returns None when the agent has no rows in the bucket-TZ month — the
    active-roster lens is deliberately NOT applied (a departed agent's
    page can still serve their closing month).
    """
    from backend.services.assessment_store import get_or_generate_month_assessment
    from backend.services.team_source import fetch_history_frame
    from backend.services.team_stats import _months_in_bucket_tz

    df = await fetch_history_frame(config)
    if df.empty:
        return None
    dm = df[_months_in_bucket_tz(df["timestamp"]) == month]
    rdf, section_cols = frame_to_render(dm, agent, config)
    if rdf.empty:
        return None
    assessment = await get_or_generate_month_assessment(
        config, agent, month, generate=generate)
    return render(rdf, agent, month, section_cols, assessment,
                  team_label=config.display_name)

File: sandy-qa/src/lib/onepager.ts
// Agent one-pager — port of backend/services/onepager.py (JulyR2R3 §3).
// Renders the print-ready monthly HTML from the SAME analytics frame the
// dashboards use, plus the PERSISTED month assessment (never generates —
// a page view must never spend a Gemini call).

import type { FrameRow } from "./historyFrame.js";
import type { TeamConfig } from "./teamConfig.js";
import { monthInBucketTz, BUCKET_TZ } from "./teamStats.js";
import { evalWindowStat } from "./coachingFacts.js";

const NAVY = "#15192D";
const BLUE = "#1A61D9";
const LIGHT_BLUE = "#E7EFFB";
const GREEN = "#28A745";
const AMBER = "#E8A317";
const RED = "#D9534F";
const GRAY = "#4A4A4A";

const esc = (s: any) =>
  String(s ?? "").replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;");

export function lastClosedMonth(now = new Date()): string {
  const ym = monthInBucketTz(now.getTime());
  const [y, m] = ym.split("-").map(Number);
  return m === 1 ? `${y - 1}-12` : `${y}-${String(m - 1).padStart(2, "0")}`;
}

const fmtG = (n: number) => String(parseFloat(n.toFixed(10))); // Python %g-ish

interface SectionStat {
  name: string;
  kind: "binary" | "numeric";
  avg: number | null;
  delta: number | null;
  na: number;
}

function mean(xs: number[]): number {
  return xs.reduce((a, b) => a + b, 0) / xs.length;
}

function sectionStat(
  name: string,
  values: (string | number | null)[]
): SectionStat {
  const strs = values.map((v) => String(v ?? "").trim());
  const isBinary = strs.some((v) => ["Yes", "Y", "No", "N"].includes(v));
  const na = strs.filter((v) => v === "Not Applicable" || v === "NA").length;
  const midpoint = Math.floor(values.length / 2);
  if (isBinary) {
    const mapped = strs
      .map((v) => (["Yes", "Y"].includes(v) ? 1 : ["No", "N"].includes(v) ? 0 : null))
      .filter((v): v is number => v !== null);
    const first = strs.slice(0, midpoint), second = strs.slice(midpoint);
    const mapHalf = (half: string[]) =>
      half
        .map((v) => (["Yes", "Y"].includes(v) ? 1 : ["No", "N"].includes(v) ? 0 : null))
        .filter((v): v is number => v !== null);
    const f = mapHalf(first), s = mapHalf(second);
    return {
      name, kind: "binary", na,
      avg: mapped.length ? Math.round(mean(mapped) * 1000) / 10 : null,
      delta: f.length && s.length ? Math.round((mean(s) - mean(f)) * 1000) / 10 : null,
    };
  }
  const nums = values
    .map((v) => (typeof v === "number" ? v : parseFloat(String(v))))
    .filter((v) => Number.isFinite(v));
  const first = nums.slice(0, Math.floor(nums.length / 2));
  const second = nums.slice(Math.floor(nums.length / 2));
  return {
    name, kind: "numeric", na,
    avg: nums.length ? Math.round(mean(nums) * 100) / 100 : null,
    delta: first.length && second.length ? Math.round((mean(second) - mean(first)) * 100) / 100 : null,
  };
}

function trendArrow(delta: number | null): [string, string] {
  if (delta === null) return ["—", GRAY];
  if (delta > 0.05) return [`▲ +${fmtG(delta)}`, GREEN];
  if (delta < -0.05) return [`▼ ${fmtG(delta)}`, RED];
  return ["→ steady", GRAY];
}

function sparkline(scores: number[], width = 560, height = 64): string {
  if (scores.length < 2) return "";
  const lo = Math.min(...scores, 60), hi = Math.max(...scores, 100);
  const span = hi - lo || 1;
  const step = width / (scores.length - 1);
  const points = scores.map((s, i) => [
    Math.round(i * step * 10) / 10,
    Math.round((height - ((s - lo) / span) * (height - 8) - 4) * 10) / 10,
  ]);
  const polyline = points.map(([x, y]) => `${x},${y}`).join(" ");
  const dots = points
    .map(([x, y]) => `<circle cx="${x}" cy="${y}" r="2.5" fill="${BLUE}"/>`)
    .join("");
  return (
    `<svg width="${width}" height="${height}" viewBox="0 0 ${width} ${height}">` +
    `<polyline points="${polyline}" fill="none" stroke="${BLUE}" stroke-width="2"/>${dots}</svg>`
  );
}

function bar(stat: SectionStat): string {
  if (stat.avg === null) return "";
  const pct = stat.kind === "binary" ? stat.avg : (stat.avg / 5) * 100;
  const color = pct >= 80 ? GREEN : pct >= 60 ? AMBER : RED;
  return `<div class="bar"><div class="fill" style="width:${Math.round(pct)}%;background:${color}"></div></div>`;
}

function assessmentHtml(assessment: any | null): string {
  if (!assessment) {
    return (
      '<div class="assessment"><strong>AI Assessment</strong><br>' +
      '<span class="muted">No persisted assessment for this window — ' +
      "the monthly export run creates one.</span></div>"
    );
  }
  const lines = (assessment.sections ?? [])
    .map((s: any) => {
      const [arrow, color] =
        s.trend === "improving" ? ["▲", GREEN] : s.trend === "declining" ? ["▼", RED] : ["→", GRAY];
      return (
        `<div class="a-sec"><span style="color:${color}">${arrow}</span> ` +
        `<strong>${esc(s.section_name)}</strong> — ${esc(s.coaching_tip)}</div>`
      );
    })
    .join("");
  const stamp = (assessment.generated_at ?? "").slice(0, 10);
  return (
    '<div class="assessment"><strong>AI Assessment</strong> ' +
    `<span class="muted">(${assessment.evaluations_included ?? "?"} evaluations` +
    `${stamp ? " · " + stamp : ""} · ${esc(assessment.rubric_version ?? "")})</span>` +
    `<p style="margin:6px 0 8px">${esc(assessment.overall_assessment)}</p>${lines}</div>`
  );
}

const MONTHS = ["January","February","March","April","May","June","July","August","September","October","November","December"];

// ── Coaching & commitments block (CoachingLoopSpec §7, CL4) ────────────────
// Sessions whose conduct date OR deadline lands in the month, with
// commitment statuses and the deterministic response readout. Renders
// nothing for months without coaching (layout unchanged — additive, like
// the assessment box).

const COMMIT_GLYPHS: Record<string, string> = {
  open: "○", met: "✓", partially_met: "◐", not_met: "✗", waived: "–",
};
const OUTCOME_COLORS: Record<string, string> = {
  met: GREEN, partially_met: AMBER, not_met: RED,
};

async function coachingBlock(
  db: D1Database,
  config: TeamConfig,
  agent: string,
  month: string
): Promise<string> {
  const roster = await db
    .prepare(
      `SELECT id FROM qa_agents WHERE team_id = ?
         AND (name = ? OR canonical_name = ?) LIMIT 1`
    )
    .bind(config.team_id, agent, agent)
    .first<any>();
  if (!roster) return "";
  const sessions = (
    await db
      .prepare(
        `SELECT * FROM qa_coachings
         WHERE team_id = ? AND agent_id = ? AND status != 'cancelled'
           AND (substr(COALESCE(completed_at, ''), 1, 7) = ?
                OR substr(COALESCE(action_plan_deadline, ''), 1, 7) = ?)
         ORDER BY COALESCE(completed_at, created_at)`
      )
      .bind(config.team_id, roster.id, month, month)
      .all<any>()
  ).results;
  if (!sessions.length) return "";

  const parts: string[] = [];
  for (const s of sessions) {
    const commits = (
      await db
        .prepare(
          "SELECT commitment, status FROM qa_coaching_commitments WHERE coaching_id = ? ORDER BY id"
        )
        .bind(s.id)
        .all<any>()
    ).results;
    let readout = "";
    if (s.completed_at) {
      const preStart = new Date(Date.parse(s.completed_at) - 30 * 86_400_000).toISOString();
      const postEnd = s.action_plan_deadline
        ? `${s.action_plan_deadline}T23:59:59Z`
        : new Date().toISOString();
      const pre = await evalWindowStat(db, config.team_id, roster.id, preStart, s.completed_at);
      const post = await evalWindowStat(db, config.team_id, roster.id, s.completed_at, postEnd);
      if (pre.n && post.n)
        readout =
          `<div class="c-readout">Response: ${pre.n} calls avg ${pre.avg} before → ` +
          `${post.n} calls avg ${post.avg} after coaching</div>`;
    }
    const outcomeBadge = s.outcome
      ? `<span class="c-outcome" style="background:${OUTCOME_COLORS[s.outcome] ?? GRAY}">` +
        `${esc(s.outcome.replace(/_/g, " "))}</span>`
      : `<span class="c-outcome" style="background:${GRAY}">confirmation due ${esc(s.action_plan_deadline ?? "—")}</span>`;
    const commitLis = commits
      .map(
        (k: any) =>
          `<div class="c-commit"><span style="color:${OUTCOME_COLORS[k.status] ?? GRAY}">` +
          `${COMMIT_GLYPHS[k.status] ?? "○"}</span> ${esc(k.commitment)}</div>`
      )
      .join("");
    parts.push(
      `<div class="c-sess">` +
        `<div><strong>${esc((s.completed_at ?? s.created_at ?? "").slice(0, 10))}</strong> ` +
        `· ${esc(s.conducted_by_role)} ${outcomeBadge}` +
        (s.action_plan_deadline ? ` <span class="muted">deadline ${esc(s.action_plan_deadline)}</span>` : "") +
        `</div>${commitLis}${readout}</div>`
    );
  }
  return (
    `<h2>Coaching &amp; commitments</h2><div class="coach-block">` +
    parts.join("") +
    `</div>`
  );
}

export async function renderMonthOnepager(
  db: D1Database,
  config: TeamConfig,
  frame: FrameRow[],
  agent: string,
  month: string,
  teamLabel: string
): Promise<string | null> {
  const rows = frame
    .filter((r) => r.agent === agent && monthInBucketTz(r.ts) === month)
    .sort((a, b) => a.ts - b.ts);
  if (!rows.length) return null;

  const overall = rows.map((r) => r.overall_score);
  const n = overall.length;
  const m = mean(overall);
  const std =
    n > 1 ? Math.sqrt(overall.reduce((a, b) => a + (b - m) ** 2, 0) / (n - 1)) : 0;
  const [yy, mm] = month.split("-").map(Number);
  const monthLabel = `${MONTHS[mm - 1]} ${yy}`;

  const stats: SectionStat[] = [];
  for (const sec of config.sections_by_number) {
    const h = sec.history_id;
    const vals = rows.map((r) =>
      h in r.num ? r.num[h] : h in r.yn ? r.yn[h] : null
    );
    if (vals.every((v) => v === null)) continue;
    stats.push(sectionStat(sec.name, vals));
  }

  // Persisted month assessment: the is_current row whose window intersects
  // the month, newest first (deviation note: Railway resolves via
  // assessment_store's month-window key; interval-intersection is the
  // D1-side equivalent for the read path).
  const monthStart = `${month}-01T00:00:00Z`;
  const monthEnd = `${month}-31T23:59:59Z`;
  const arow = await db
    .prepare(
      `SELECT a.*, ag.name AS agent_name FROM qa_assessments a
       JOIN qa_agents ag ON ag.id = a.agent_id
       WHERE a.team_id = ? AND a.is_current = 1
         AND (ag.name = ? OR ag.canonical_name = ?)
         AND a.range_start_at <= ? AND a.range_end_at >= ?
       ORDER BY a.generated_at DESC LIMIT 1`
    )
    .bind(config.team_id, agent, agent, monthEnd, monthStart)
    .first<any>();
  let assessment: any = null;
  if (arow) {
    const secs = await db
      .prepare(
        "SELECT section_name, trend, coaching_tip FROM qa_assessment_sections WHERE assessment_id = ? ORDER BY section_number"
      )
      .bind(arow.id)
      .all<any>();
    assessment = { ...arow, sections: secs.results };
  }

  const tableRows = stats
    .map((s) => {
      const [arrow, color] = trendArrow(s.delta);
      const avgCell =
        s.avg === null ? "—" : s.kind === "binary" ? `${fmtG(s.avg)}% Yes` : `${fmtG(s.avg)} / 5`;
      const naCell = s.na ? `${s.na} NA` : "";
      return (
        `<tr><td>${esc(s.name)}</td><td class="num">${avgCell}</td>` +
        `<td>${bar(s)}</td><td class="num" style="color:${color}">${arrow}</td>` +
        `<td class="num muted">${naCell}</td></tr>`
      );
    })
    .join("\n");

  const genStamp = new Intl.DateTimeFormat("sv-SE", {
    timeZone: BUCKET_TZ, dateStyle: "short", timeStyle: "short",
  }).format(new Date());

  const coachingHtml = await coachingBlock(db, config, agent, month);

  return `<!DOCTYPE html>
<html><head><meta charset="utf-8"><title>${esc(agent)} — QA ${monthLabel}</title>
<style>
  @page { size: letter portrait; margin: 0.55in; }
  * { box-sizing: border-box; margin: 0; }
  body { font: 13px/1.45 -apple-system, "Segoe UI", Helvetica, Arial, sans-serif;
         color: ${GRAY}; max-width: 7.4in; margin: 0 auto; padding: 24px 8px; }
  header { background: ${NAVY}; color: #fff; border-radius: 10px;
            padding: 18px 22px; display: flex; justify-content: space-between;
            align-items: baseline; }
  header h1 { font-size: 21px; font-weight: 650; }
  header .brand { color: ${LIGHT_BLUE}; font-size: 12px; letter-spacing: 0.12em;
                   text-transform: uppercase; }
  .cards { display: grid; grid-template-columns: repeat(4, 1fr); gap: 10px;
            margin: 14px 0; }
  .card { background: ${LIGHT_BLUE}; border-radius: 8px; padding: 12px 14px; }
  .card .v { font-size: 26px; font-weight: 700; color: ${NAVY};
              font-variant-numeric: tabular-nums; }
  .card .l { font-size: 10.5px; text-transform: uppercase; letter-spacing: 0.08em; }
  h2 { font-size: 12px; color: ${NAVY}; text-transform: uppercase;
        letter-spacing: 0.1em; margin: 18px 0 8px; }
  table { width: 100%; border-collapse: collapse; }
  td { padding: 5px 8px; border-bottom: 1px solid #eceff4; }
  td.num { text-align: right; white-space: nowrap;
            font-variant-numeric: tabular-nums; }
  .muted { color: #9aa3af; font-size: 11px; }
  .bar { width: 130px; height: 7px; background: #e8ebf0; border-radius: 4px; }
  .fill { height: 100%; border-radius: 4px; }
  .assessment { border: 1.5px dashed ${BLUE}; border-radius: 10px; padding: 14px 16px;
                 margin-top: 16px; color: ${NAVY}; background: #fbfcff; }
  .a-sec { font-size: 11.5px; margin: 2px 0; }
  .coach-block { border: 1px solid #e8ebf0; border-radius: 10px; padding: 12px 14px; }
  .c-sess { margin-bottom: 10px; font-size: 12px; }
  .c-sess:last-child { margin-bottom: 0; }
  .c-outcome { color: #fff; font-size: 10px; padding: 1px 8px; border-radius: 8px;
               text-transform: uppercase; letter-spacing: 0.05em; }
  .c-commit { font-size: 11.5px; margin: 3px 0 0 8px; }
  .c-readout { font-size: 11px; color: #9aa3af; margin-top: 4px; }
  footer { margin-top: 18px; font-size: 10.5px; color: #9aa3af; }
</style></head><body>
<header>
  <div><h1>${esc(agent)}</h1>
  <div style="color:${LIGHT_BLUE}">${esc(teamLabel)} — QA Performance</div></div>
  <div style="text-align:right"><div class="brand">Landing · QA</div>
  <div style="font-size:15px;font-weight:600">${monthLabel}</div></div>
</header>

<div class="cards">
  <div class="card"><div class="v">${m.toFixed(1)}</div><div class="l">Avg overall score</div></div>
  <div class="card"><div class="v">${n}</div><div class="l">Calls evaluated</div></div>
  <div class="card"><div class="v">${std.toFixed(1)}</div><div class="l">Std deviation</div></div>
  <div class="card"><div class="v">${Math.min(...overall).toFixed(0)}–${Math.max(...overall).toFixed(0)}</div><div class="l">Score range</div></div>
</div>

<h2>Score progression — ${monthLabel}</h2>
${sparkline(overall)}

<h2>Per-section performance</h2>
<table><tbody>
${tableRows}
</tbody></table>

${coachingHtml}

${assessmentHtml(assessment)}

<footer>Generated ${genStamp} · Source: qa_evaluations
(${n} finalized evaluations, ${monthLabel}) · Trend = second-half vs first-half average
· Scores computed by the Landing QA engine</footer>
</body></html>`;
}
